enhancements/machine-config/machine-config-node.md

```diff
@@ -0,0 +1,340 @@
+
+## Motivation
+
+The MCO manages node upgrades but since we do not own the object or store much of their data in other ways, much of what occurs during an upgrade
+is simply a black box operation that we currently report as "Updating" or "Updated". Users can debug into a specific node or look into the node spec for some of this information
+but most of it simply lives in the MCO code rather than in data structures. We want to put these abstract "phases" of node operations as triggered by the MCO into a concrete data structure. 
+
+### User Stories
+
+* As a cluster admin, during upgrades I would like to monitor the progress of individual node upgrades so that, if one of the nodes sticks during an upgrade, I can easily see where the node got stuck and why.
+* As an OpenShift developer, I would like a place to observe the MCO's view of a Node so that I can easily identify if the MCO is related to an upgrade failure in CI.
+
+### Goals
+
+* Make a MachineConfigNode type that succinctly holds the upgrade data.
+* Have API load be as minimal as possible but augment the proper objects as needed.
+* aggregate as much MCO related data into easily accessible places as possible.
+
+### Non-Goals
+
+* Make existing/new API objects difficult to understand.
+
+## Proposal
+Create a Datatype for tracking Node Upgrade Progression in the MCO as well as Operator Component Progression in the MCO.
+
+Create a mechanism inside of the MCO to update the new MachineConfigNode API type
+with data about a Node's progress during an upgrade and any errors that occur during those
+processes.
+
+The data created by this MachineConfigNode would look like the following:
+
+```console
+$ oc get machineconfignodes
+NAME                          UPDATED   UPDATEPREPARING   UPDATEINPROGRESS   UPDATEPOSTACTION   UPDATECOMPLETING   RESUMING
+ip-10-0-12-194.ec2.internal   True      False             False              False              False              False
+ip-10-0-17-102.ec2.internal   True      False             False              False              False              False
+ip-10-0-2-232.ec2.internal    True      False             False              False              False              False
+ip-10-0-59-251.ec2.internal   True      False             False              False              False              False
+ip-10-0-59-56.ec2.internal    True      False             False              False              False              False
+ip-10-0-6-214.ec2.internal    True      False             False              False              False              False
+```
+
+as well as 
+
+```console
+$ oc get machineconfignodes -o wide
+NAME                          UPDATED   UPDATEPREPARING   UPDATEINPROGRESS   UPDATEPOSTACTION   UPDATECOMPLETING   RESUMING   COMPARINGMACHINECONFIGS   UPDATINGFILESANDOS   DRAININGNODE   CORDONINGNODE   REBOOTINGNODE   RELOADINGCRIO
```

Comment:
RELOADINGCRIO is one of the post-update action. Maybe this can be extra detail as part of UPDATEPOSTACTION? This will also keep external api field independent of what all services MCO does live reload/restart and any future update do around it.

Reply:
As we have it today, would `UPDATEPOSTACTION` and `RELOADINGCRIO` both be true at the same time?

Reply:
It's not clear to me how you'd deliver all these output columns in `oc get machineconfignodes -o wide` if they are being fed by `status.conditions` like your later example (which only has an `Updated` entry).  Would the MachineConfigNode controller try to populate all of these conditions on all resources, with entries like `UpdatePostAction=False` with a `reason` like `WaitingOnPreviousCondition` and a `message` like `Update post-actions are not attempted until...`?

Reply:
My expectation is that what @wking has said he is what should happen, but reading some of the responses to comments on my last review, I think that's not the way the MCO currently does conditions, and we need to make sure there's some alignment

Reply:
If only one state and one substate can be true at a given time maybe shorten it to a phase when pretty printing? I feel like that'd be easiest to consume when trying to debug why something has gone wrong. Not sure whether the underlying API should follow this or remain as suggested, just thinking about the user experience.

```console
$ oc get machineconfignodes
NAME    UPDATED   PHASE
node-1  False     UpdateInProgress: CordoningNode
node-2  True      Idle
$ oc get machineconfignodes
NAME    UPDATED   PHASE
node-1  False     UpdateInProgress: DrainingNode 
node-2  True      Idle
$ oc get machineconfignodes
NAME    UPDATED   PHASE
node-1  False     UpdateInProgress: DrainingNode -- retry, timed out waiting on pods accounting:pod-ABCDE to drain
node-2  True      Idle
$ oc delete pod -n accounting pod-ABCDE -f
$ oc get machineconfignodes
NAME    UPDATED   PHASE
node-1  False     UpdateInProgress: UpdatingFilesAndOS
node-2  True      Idle

```

How do MAPI and CAPI handle machine phase statemachine on the backend?

Reply:
@sdodson I am not sure if we can have something like:

```
$ oc get machineconfignodes
NAME    UPDATED   PHASE
node-1  False     UpdateInProgress: CordoningNode
node-2  True      Idle
```

and then 

```
oc get machineconfignodes -o wide
NAME                          UPDATED   UPDATEPREPARING   UPDATEINPROGRESS...
```
I don't think you can get rid of a column between the pretty print and -o wide. so it would instead look like

```
$ oc get machineconfignodes - o wide
NAME    UPDATED   PHASE                                                       UPDATED   UPDATEPREPARING   UPDATEINPROGRESS...
node-1  False           UpdateInProgress: CordoningNode
node-2  True           Idle
```
This might be a little weird, but I am open to it potentially.  So phase would always have to exist which wouldn't make sense with -o wide. @JoelSpeed thoughts here? I believe that we decided on this design due to the fact that it conforms with what other operators do.

If there is a way to aggregate this to the console instead, I would be open to that! But not sure how that works in this scenario... would that be an acceptable alternative?

Reply:
> How do MAPI and CAPI handle machine phase statemachine on the backend?

I'm not sure MAPI/CAPI are shining examples of the best way to do this tbh, but, we have a phase field which the controller updates with the current phase, and then the printer column just prints from that value.

I guess we could add a phase enum as long as that is kept up to date, will need to have a think.
I do also wonder if the custom printer column logic could pick this out with some JQ, will have to look into that.

I think a succinct phase could be useful though so we should work this out.

Reply:
> It's not clear to me how you'd deliver all these output columns in `oc get machineconfignodes -o wide` if they are being fed by `status.conditions` like your later example (which only has an `Updated` entry). Would the MachineConfigNode controller try to populate all of these conditions on all resources, with entries like `UpdatePostAction=False` with a `reason` like `WaitingOnPreviousCondition` and a `message` like `Update post-actions are not attempted until...`?

When transitioning to a false condition, I am going to modify the message to be in the negative rather than keep the old message. I can also work on the wording for when the condition had not occurred yet.